### PR TITLE
BUGFIX: Prevent unsigned integer underflow in DocumentUriPathProjection when duplicate SubtreeWasUntagged events reach live

### PR DESCRIPTION
**Pull request by skurfuerst**
Before decrementing the disabled/removed counter, check whether the node was actually explicitly tagged by comparing its counter with the parent node's counter. If the node's level is not greater than the parent's, the untag is a no-op — preventing 0 - 1 on the UNSIGNED column.

Resolves: #5778

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the correct branch:
  - If it's a bugfix, use the [lowest maintained branch which has the bug](https://www.neos.io/features/release-roadmap.html)
  - If it's a non-breaking feature, use the branch of the next version (might be either minor or major)
  - If it's a breaking feature it should typically go into the next major version
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php [hunk 1]
```diff
@@ -453,6 +453,24 @@ private function whenSubtreeWasUntagged(SubtreeWasUntagged $event): void
...
+                $nodeTagLevel = $event->tag->equals(NeosSubtreeTag::disabled()) ? $node->getDisableLevel() : $node->getRemovedLevel();
+                $parentNode = $this->tryGetNode(fn () => $this->documentUriPathFinder->getParentNode($node));
+                $parentTagLevel = 0;
+                if ($parentNode !== null) {
+                    $parentTagLevel = $event->tag === NeosSubtreeTag::disabled() ? $parentNode->getDisableLevel() : $parentNode->getRemovedLevel();
+                }
+                if ($nodeTagLevel <= $parentTagLevel) {
+                    // Node was not explicitly tagged (its counter matches [or is below - should never happen] the parent's level).
+                    // Decrementing might cause an unsigned integer underflow. This can happen when
+                    // a duplicate SubtreeWasUntagged event reaches the live stream (e.g. due to
```
Comment (kitsunet) [inline Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php:467]: We need to investigate why this can happen though, because (as per discussion in slack) it should not, and was always a big no go. So we should avoid it on event creation/publishing. Unless we are willing to change that and allow duplicated taggings as the second (and more) are essentially no-ops.

**Review by kitsunet — APPROVED** [review]

**Comment by skurfuerst** [comment]
FYI: Discussion between @kitsunet and me on Slack at https://neos-project.slack.com/archives/C04PYL8H3/p1776086953688529 :

@kitsunet 
> the problem is, I don't remember what our last stance here was. I know I argued repeatedly that two concurrent disable: false are no problem at all, because the result is the same if it is applied once or twice, but I was told again and again that tthis should not be allowed to happen. So  I wonder where that changed. Or did I forget that we actually changed this and allow this kind of even duplicates now

@skurfuerst :

> well I don't know, but for some reason it happened in production :wink:
> although it should not by reading the code ^^
> I try to fix all issues I am hitting with Neos 9 in prod right now ^^

@kitsunet 
> Right, but then I think we need to look at that, because it was made very clear to  me that we cannot have these  kind of duplicate events even if it doesn't  matter
> I am generally fine with the fix though

@skurfuerst 
> "we cannot have these kind of duplicate events even if it doesn't matter" -> Just spent like 2 hours on trying to find the root cause, but was not able to ^^

**Comment by mhsdesign** [comment]
Thanks for bringing this up again. The discussion is age old.

My pr with beta16 https://github.com/neos/neos-development-collection/pull/5485/ removed the condition under following argumentation:

> The `isNodeExplicitlyDisabled` check was dropped as discussed for the case of just updating the disabled state.
Previously the projection could handle a case where a node was disabled, or re-enabled after already being in that state.
While partly controversial, we agreed that this behaviour is not valid and is to be caught by constraint checks as implemented by https://github.com/neos/neos-development-collection/pull/5357
> If however we decide to allow to disable a node twice - the same as setting a property to the same value again - we must adjust the projection to be robust for both the disabled and removed state, and then introduce test using the actual command api instead of hackily publishing events.

And my last stand as we discussed this in slack a year ago was:  https://neos-project.slack.com/archives/C04PYL8H3/p1747947129227789

> We had the discussion a LOT of times if we should build the projections to be forgiving (sometimes they are by design - like the content graph does not die in the disabled case) or if they should always be really strict and reject all invalid events ...
having the projections always be forgiving is also impossible - e.g. you cannot create a node without parent - otherwise its orphaned - and then making it sometimes forgiving is also just a vague definition. [...]
> 
> But the event stream must always be correct ... well now it turns out to not be the case? Lets see - i say good for now that we notice that something is odd instead of letting it fly by? [...]

I hoped that the events were from old beta times at that time but if we still have the duplication problem something is conceptional wrong in the locking and we need to find out. You wrote

> We observed two SubtreeWasUntagged events (originating from two separate UntagSubtree commands) for the same node aggregate on the same workspace/content stream, recorded seconds apart. Both were original commands — not rebased copies. When published to live, both untag events were applied, causing the counter to be decremented twice for a single tag.

So my question is how can we create a test to reproduce that? Possibly a parallel test? Ill try, but even that is impossible due to version checks. And during rebase or publish we reject duplicate disables.

**Comment by skurfuerst** [comment]
@mhsdesign I tried very hard to build a test case, but was not able to reproduce this. The events/commands were 7 seconds apart, so too long to be "extremely racey", but still triggered by the user somehow.

I also checked the code intensely, and have no idea how this might happen.

**Could we merge this fix please, to ensure the users are not hit by this stop-the-world-problem; and additionally try to understand it further? :-)**

All the best,
Sebastian

**Comment by kitsunet** [comment]
I don't see the harm in merging this _now_ ? So lets do it...

**Review by mhsdesign — APPROVED** [review]
I tried to build a parallel test to reproduce this and find any underlying problems so far. Bernhard suggested to use a line of death to slow down tagging. But i cannot reproduce the behaviour still.

The test is here https://github.com/neos/neos-development-collection/pull/5788 maybe you have an idea. But for now lets reintroduce this patch.

**Comment by skurfuerst** [comment]
Thanks so much :):)

**Comment by mhsdesign** [comment]
Coming back to it - and because it's your birthday - i have to admit you were fully right with these adjustments.

But the reasoning _why_ we need the change is different.

I was now able to reproduce the problems quite simple just by taking dimensions into account. See various test cases i came up with: https://github.com/neos/neos-development-collection/pull/5901 

E.g. if a user would have disabled first a node in the specialisation and after that also in the generalisation it crashed before. Its not a race condition and also its not the _same_ event twice.

Good thing also that you did recreate the change and not just reverted my revert from beta16:
Because as concluded above https://github.com/neos/neos-development-collection/pull/5779#issuecomment-4243116784 beta16 had some kind of underflow protection.
The underflow protection it had was just only build early on to handle two succedent `DisableNodeAggregate` commands ([was legal in early ages it seems](https://github.com/neos/neos-development-collection/commit/6f12827b875448a73d4b2f9a4ec2a9b55902eea7#diff-39d892d840ec8a7261292d943f59f15b8667887665b772d944098cc433e309eaR172)) and it was also only tested in the zero dimensional case. The pre beta16 implementation did not satisfy my tests as there was an early return instead of `continue` - which meant all other dimensions were skipped.

https://github.com/neos/neos-development-collection/blob/397737cfaa6f8e6057050369971ab57d90a392d4/Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php#L490-L493
